lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2490,25 +2490,41 @@ namespace {
         // that we add this to the parent enum (in the "__ObjC" module) and not
         // to the extension.
         auto parentNS = cast<clang::NamespaceDecl>(decl->getParent());
-        auto parent = Impl.importDecl(parentNS, getVersion());
+        auto parent =
+            Impl.importDecl(parentNS, getVersion(), /*UseCanonicalDecl*/ false);
         // Sometimes when the parent namespace is imported, this namespace
         // also gets imported. If that's the case, then the parent namespace
         // will be an enum (because it was able to be fully imported) in which
         // case we need to bail here.
-        auto cachedResult =
-            Impl.ImportedDecls.find({decl->getCanonicalDecl(), getVersion()});
+        auto cachedResult = Impl.ImportedDecls.find({decl, getVersion()});
         if (cachedResult != Impl.ImportedDecls.end())
           return cachedResult->second;
         dc = cast<ExtensionDecl>(parent)
                  ->getExtendedType()
                  ->getEnumOrBoundGenericEnum();
       }
-      auto *enumDecl = Impl.createDeclWithClangNode<EnumDecl>(
-          decl, AccessLevel::Public, loc,
-          importedName.getDeclName().getBaseIdentifier(),
-          Impl.importSourceLoc(decl->getLocation()), None, nullptr, dc);
-      if (isa<clang::NamespaceDecl>(decl->getParent()))
-        cast<EnumDecl>(dc)->addMember(enumDecl);
+
+      EnumDecl *enumDecl = nullptr;
+      // Try to find an already created enum for this namespace.
+      for (auto redecl : decl->redecls()) {
+        if (Impl.ImportedDecls.find({redecl, getVersion()}) !=
+            Impl.ImportedDecls.end()) {
+          auto ext = dyn_cast<ExtensionDecl>(
+              Impl.ImportedDecls[{redecl, getVersion()}]);
+          enumDecl = dyn_cast<EnumDecl>(ext->getExtendedNominal());
```

Comment:
It's a map, so it probably doesn't matter too much, but we can avoid looking this up twice. Also, this should _always_ be an extension decl, and the extension should _always_ be for an enum. If that's not the case, something went seriously wrong (so, we can just use `cast` which will assert that's true). 

```suggestion
        auto ext = Impl.ImportedDecls.find({redecl, getVersion()});
        if (ext != Impl.ImportedDecls.end()) {
          enumDecl = cast<EnumDecl>(cast<ExtensionDecl>(ext)->getExtendedNominal());
```

Reply:
Thanks, I had to change it slightly to make it compile (`ext` to `ext.second`) so applied it manually.